### PR TITLE
Feature/slonik plugin

### PR DESCRIPTION
**Pull request by megquirk**
### Summary

- Moved [Slonik Plugin](https://github.com/autotelic/harpocrates/tree/main/plugins/fastify-slonik) from Harpocrates repo.

### Test Plan

- Clone and install `npm i`
- To confirm the plugin is working, clone and install Harpocrates project from [this repo](https://github.com/autotelic/harpocrates)
  - `cd` into the project directory and edit location of fastifySlonik in app.js
  ```js
  ...
  const fastifySlonik = require('<path to this plugin, i.e. ../fastify-slonik>')
  ...
  ```
  - Create .envrc `cp .envrc.example .envrc`
  - Update .envrc with public ip
    ```sh
    $ ifconfig -u | grep 'inet ' | grep -v 127.0.0.1 | cut -d\  -f2 | head -1
    # <your-public-ip> <-- Copy this
    ```
  - Update .envrc with copied public-ip `export PUBLIC_IP=<your-public-ip>`
  - Run the following commands to create and run [docker](https://www.docker.com/) container `docker-compose up`
  - Seed and migrate database
  ```sh
  $ cp test_seeds/1605303453216_seeds.js migrations/
  $ npm run migrate up
  ```
  - Run in development `npm run dev`
- If plugin is working correctly, GET request to `localhost:3000/users/1` should return
```js
{
userId: 1,
githubUsername: "Ed",
emailAddress: "Ed@autotelic.com",
firstLogin: 1610521099724,
mostRecentLogin: 1610521099724
}
```

### To Do
- index.test.js is failing, need to create example.js instead of requiring harpocrates app to test

### CODE REVIEW

**Comment by eadmundo** [comment]
This is looking great! What I would do to make it easier to review & test is to create an accompanying PR in the harpocrates repo which uses this plugin - you can use `autotelic/fastify-slonik#feature/slonik-plugin` as the package name in the dependencies. That PR can then be updated once we release this to npm.

**Review by jkirkpatrick24 — APPROVED** [review]
LGTM - Lets just remove the empty test.js file 👍 Great stuff!

File: index.js [hunk 1]
```diff
@@ -1,13 +1,27 @@
...
+  const db = {
+    connect: pool.connect.bind(pool),
+    pool: pool
+  }
+
+  fastify.decorate('slonik', db)
+  fastify.decorate('sql', sql)
+}
+
+module.exports = fastifyPlugin(fastifySlonik)
```
Comment (jkirkpatrick24) [inline index.js]: With fresh eyes, I realize we should have a name for this plugin. See the [fastify docs here](https://github.com/fastify/fastify-plugin#name)

The TLDR is it allows us to do dependency checks, when loading other plugins. (Make sure this plugin exists and has been registered before we try to use it)

**Comment by megquirk** [comment]
I think I've taken care of most of the issues:
- README updated
- tests are all passing
- plugin has been named

**Review by jkirkpatrick24 — COMMENTED** [review]
This is looking really really good. Docs are great and the tests cover what we want. I think the last piece is to setup a postgres instance with docker and to have the tests connect to that docker instance.

We can cargo-cult a lot from `fastify-postgres`

[running postgres on docker](https://github.com/fastify/fastify-postgres/blob/master/package.json#L13)

This will allow anyone to run the tests successfully, as well as allow us to configure tests to run in a CI environment (which we are going to want for sure)

**Comment by matthew-charles-chan** [comment]
> This is looking really really good. Docs are great and the tests cover what we want. I think the last piece is to setup a postgres instance with docker and to have the tests connect to that docker instance.
> 
> We can cargo-cult a lot from `fastify-postgres`
> 
> [running postgres on docker](https://github.com/fastify/fastify-postgres/blob/master/package.json#L13)
> 
> This will allow anyone to run the tests successfully, as well as allow us to configure tests to run in a CI environment (which we are going to want for sure)

We can take reference from the `fastify-postgres` [Development and Testing docs](https://github.com/fastify/fastify-postgres#development-and-testing) to add to our docs as well

**Comment by megquirk** [comment]
Updated 

> This is looking really really good. Docs are great and the tests cover what we want. I think the last piece is to setup a postgres instance with docker and to have the tests connect to that docker instance.
> 
> We can cargo-cult a lot from `fastify-postgres`
> 
> [running postgres on docker](https://github.com/fastify/fastify-postgres/blob/master/package.json#L13)
> 
> This will allow anyone to run the tests successfully, as well as allow us to configure tests to run in a CI environment (which we are going to want for sure)



> > This is looking really really good. Docs are great and the tests cover what we want. I think the last piece is to setup a postgres instance with docker and to have the tests connect to that docker instance.
> > We can cargo-cult a lot from `fastify-postgres`
> > [running postgres on docker](https://github.com/fastify/fastify-postgres/blob/master/package.json#L13)
> > This will allow anyone to run the tests successfully, as well as allow us to configure tests to run in a CI environment (which we are going to want for sure)
> 
> We can take reference from the `fastify-postgres` [Development and Testing docs](https://github.com/fastify/fastify-postgres#development-and-testing) to add to our docs as well

I've updated the scripts and docs to setup  a postgres instance with docker.

**Review by jkirkpatrick24 — APPROVED** [review]
LGTM! Great  job. 

There is just one empty pesky test.js file we should remove but then squash and merge away!

**Pull request merged**